ui/vty/src/Ariadne/UI/Vty/Widget.hs

```diff
@@ -0,0 +1,394 @@
+instance B.Named (WidgetInfo s p) WidgetName where
+  getName = widgetName
+
+data Widget p = forall s. Widget (WidgetInfo s p)
+
+instance B.Named (Widget p) WidgetName where
+  getName (Widget WidgetInfo{..}) = widgetName
+
+type WidgetInitM s p = State (WidgetInfo s p) ()
+
+type WidgetDrawM s p a = ReaderT (WidgetInfo s p) (Reader p) a
+
+type WidgetEventM s p a = StateT s (StateT (WidgetInfo s p) (StateT p (B.EventM WidgetName))) a
+
+makeLensesWith postfixLFields ''WidgetInfo
+
+----------------------------------------------------------------------------
+-- Widget initialization
+----------------------------------------------------------------------------
+
+initWidget :: WidgetInitM s p -> Widget p
+initWidget action =
+  Widget $ execState action WidgetInfo
+    { widgetName = []
+    , widgetState = error "State not defined for this widget"
+    , widgetChildren = Map.empty
+    , widgetFocusList = []
+    , widgetEventHandlers = Map.empty
+    , widgetEventQueue = []
+    , widgetEventSend = return . const ()
+    , widgetDraw = return . const B.emptyWidget
+    , widgetDrawWithFocused = \_ s -> do
+        draw <- view widgetDrawL
+        draw s
+    , widgetDrawWithFocus = \focus s -> do
+        draw <- view widgetDrawWithFocusedL
+        name <- view widgetNameL
+        draw (name == focus) s
+    , widgetHandleEditKey = return . const WidgetEventNotHandled
+    , widgetHandleKey = return . const WidgetEventNotHandled
+    , widgetHandlePaste = return . const WidgetEventNotHandled
+    , widgetHandleMouseDown = return . const WidgetEventNotHandled
+    , widgetHandleScroll = return . const WidgetEventNotHandled
+    , widgetHandleEvent = return . const ()
+    }
+
+setWidgetState :: s -> WidgetInitM s p
+setWidgetState = assign widgetStateL
+
+addWidgetChild :: WidgetNamePart -> Widget (WidgetInfo s p) -> WidgetInitM s p
+addWidgetChild namePart (Widget child) = do
+    let child' = child{ widgetEventSend = \event -> lift . lift $ widgetEventQueueL %= ((namePart, event):) }
+    widgetChildrenL %= Map.insert namePart (rename $ Widget child')
+  where
+    rename :: Widget p -> Widget p
+    rename (Widget widget@WidgetInfo{..}) =
+      Widget widget
+        { widgetName = namePart : widgetName
+        , widgetChildren = Map.map rename widgetChildren
+        }
+
+setWidgetFocusList :: MonadState (WidgetInfo s p) m => [WidgetNamePart] -> m ()
+setWidgetFocusList = assign widgetFocusListL
+
+addWidgetEventHandler :: MonadState (WidgetInfo s p) m => WidgetNamePart -> (WidgetEvent -> WidgetEventM s p ()) -> m ()
+addWidgetEventHandler namePart handler = do
+  widgetEventHandlersL %= Map.insert namePart handler
+
+setWidgetDraw :: MonadState (WidgetInfo s p) m => (s -> WidgetDrawM s p (B.Widget WidgetName)) -> m ()
+setWidgetDraw = assign widgetDrawL
+
+setWidgetDrawWithFocused :: MonadState (WidgetInfo s p) m => (Bool -> s -> WidgetDrawM s p (B.Widget WidgetName)) -> m ()
+setWidgetDrawWithFocused = assign widgetDrawWithFocusedL
```

Comment:
Is there really no way to avoid this repetitive code?

Reply:
We could probably just expose lenses, but I'm not sure yet.

Reply:
I would rather leave it as-is, because this way we can easily change inner workings without rewriting all widget constructors.